src/mnode/mnode-pastel.cpp

```diff
@@ -1442,3 +1442,121 @@ bool CTakeDownTicket::FindTicketInDb(const std::string& key, CTakeDownTicket& ti
 {
     return false;
 }
+
+
+// CChangeUsernameTicket ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+CChangeUsernameTicket CChangeUsernameTicket::Create(std::string _pastelID, std::string username, const SecureString& strKeyPass)
+{
+    CChangeUsernameTicket ticket(std::move(_pastelID));
+    
+    ticket.username = std::move(username);
+
+    CChangeUsernameTicket _ticket;
+    _ticket.pastelID = _pastelID;
+    // Check if PastelID already have a username on the blockchain. 
+    if (!masterNodeCtrl.masternodeTickets.FindTicketBySecondaryKey(_ticket)) {
+        // IF PastelID has no Username yet, the fee is 100 PSL
+        ticket.fee = 100;
+    } else {
+        // IF PastelID changed Username before, fee should 5000
+        ticket.fee = 5000;
+    }
+
+    ticket.GenerateTimestamp();
+    
+    std::string strTicket = ticket.ToStr();
+    ticket.signature = CPastelID::Sign(reinterpret_cast<const unsigned char*>(strTicket.c_str()), strTicket.size(), ticket.pastelID, strKeyPass);
+    
+    return ticket;
+}
+
+bool CChangeUsernameTicket::FindTicketInDb(const std::string& key, CChangeUsernameTicket& ticket)
+{
+    ticket.username = key;
+    return masterNodeCtrl.masternodeTickets.FindTicket(ticket);
+}
+
+
+std::string CChangeUsernameTicket::ToJSON() const noexcept
+{
+    json jsonObj;
+    jsonObj = {
+        {"txid", m_txid},
+        {"height", m_nBlock},
+        {"ticket", {
+            {"type", GetTicketName()}, 
+            {"pastelID", pastelID},
+            {"username", username}, 
+            {"fee", fee}, 
+            {"signature", ed_crypto::Hex_Encode(signature.data(), signature.size())}}}};
+
+    return jsonObj.dump(4);
+}
+
+std::string CChangeUsernameTicket::ToStr() const noexcept
+{
+    std::stringstream ss;
+    ss << pastelID;
+    ss << username;
+    ss << fee;
+    ss << m_nTimestamp;
+    return ss.str();
+}
+
+bool CChangeUsernameTicket::IsValid(std::string& errRet, bool preReg, int depth) const
+{
+    unsigned int chainHeight = 0;
+    {
+        LOCK(cs_main);
+        chainHeight = static_cast<unsigned int>(chainActive.Height()) + 1;
+    }
+
+    // A. Check if the username is already registered on the blockchain.
+    if (masterNodeCtrl.masternodeTickets.CheckTicketExist(*this)) {
+        errRet = strprintf("This Username is already registered in blockchain [Username = %s]", username);
+        return false;
+    }
+
+    // B Verify signature
+    // We will check that it is the correct PastelID
+    std::string strThisTicket = ToStr();
+    if (!CPastelID::Verify(reinterpret_cast<const unsigned char *>(strThisTicket.c_str()), strThisTicket.size(),
+                           signature.data(), signature.size(),
+                           pastelID))
+    {
+        errRet = strprintf("%s ticket's signature is invalid. PastelID - [%s]", "Username", pastelID);
```

Comment:
Can use GetTicketDescription(TicketID::Username) instead of "Username"

Reply:
fixed